src/OrchardCore.Modules/OrchardCore.Media/Services/MediaFileStoreResolverMiddleware.cs

```diff
@@ -0,0 +1,144 @@
+
+        private readonly RequestDelegate _next;
+        private readonly ILogger<MediaFileStoreResolverMiddleware> _logger;
+        private readonly IMediaFileStoreCache _mediaFileStoreCache;
+        private readonly IMediaFileStore _mediaFileStore;
+
+        private readonly PathString _assetsRequestPath;
+        private readonly HashSet<string> _allowedFileExtensions;
+
+        public MediaFileStoreResolverMiddleware(
+            RequestDelegate next,
+            ILogger<MediaFileStoreResolverMiddleware> logger,
+            IMediaFileStoreCache mediaFileStoreCache,
+            IMediaFileStore mediaFileStore,
+            IOptions<MediaOptions> mediaOptions
+            )
+        {
+            _next = next;
+            _logger = logger;
+            _mediaFileStoreCache = mediaFileStoreCache;
+            _mediaFileStore = mediaFileStore;
+
+            _assetsRequestPath = mediaOptions.Value.AssetsRequestPath;
+            _allowedFileExtensions = mediaOptions.Value.AllowedFileExtensions;
+        }
+
+        /// <summary>
+        /// Processes a request to determine if it matches a known file from the media file store, and cache it locally.
+        /// </summary>
+        /// <param name="context"></param>
+        public async Task Invoke(HttpContext context)
+        {
+            //TODO for 3.0 and endpoint routing, validate it is not an endpoint
+
+            // Support only Head requests or Get Requests.
+            if (!HttpMethods.IsGet(context.Request.Method) && !HttpMethods.IsHead(context.Request.Method))
+            {
+                await _next(context);
+                return;
+            }
+
+            var validatePath = context.Request.Path.StartsWithSegments(_assetsRequestPath, StringComparison.OrdinalIgnoreCase, out var subPath);
+            if (!validatePath)
+            {
+                _logger.LogDebug("Request path {Path} does not match the assets request path {RequestPath}", subPath, _assetsRequestPath);
+                await _next(context);
+                return;
+            }
+
+            // This will not cache the file if the file extension is not supported.
+            var fileExtension = GetExtension(subPath);
+            if (!_allowedFileExtensions.Contains(fileExtension))
+            {
+                _logger.LogDebug("File extension not supported for request path {Path}", subPath);
+                await _next(context);
+                return;
+            }
+
+            var isFileCached = await _mediaFileStoreCache.IsCachedAsync(subPath);
+            if (isFileCached)
+            {
+                // When multiple requests occur for the same file the download 
+                // may already be in progress so we wait for it to complete.
+                if (Workers.TryGetValue(subPath, out var writeTask))
+                {
+                    await writeTask.Value;
+                }
+
```

Comment:
Oh yes, maybe because `IsCachedAsync` returns true as soon as we do `File.Create` in `SetCacheAsync` and before having completed `stream.CopyToAsync`, so we have to wait that the the lazy value has completed, right?

Reply:
Yes exactly, and I used Lazy (maybe Sebastien still change his mind) because it helps handling task creation without a race, knowing that these tasks are heavy and you don't want to create two, and discard one because of a race. (thinking of the `DynamicCacheTagHelper` `Workers` `TryGetValue` and `TryAdd`.

But just discovering these for the first time so...

Reply:
@jtkech I think the locations where you used Lazy and it wasn't necessary had nothing to do with `Lazy<Task>`